File: 0473-matchsticks-to-square/0473-matchsticks-to-square.js
/**
 * @param {number[]} matchsticks
 * @return {boolean}
 */
var makesquare = function (matchsticks) {
    let sum = matchsticks.reduce((a, v) => a + v)
    if (sum % 4 !== 0) return false
    let target = sum / 4

    let array = new Array(4).fill(0)
    matchsticks = matchsticks.sort((a, b) => b - a)
    var again = function (start) {
        if (start == matchsticks.length) {
            return true
        }

        for (let i = 0; i < 4; i++) {
            if (matchsticks[start] + array[i] <= target) {
                array[i] = matchsticks[start] + array[i]
                if (again(start + 1)) return true
                array[i] = array[i] - matchsticks[start]
            }
        }
        return false
    }
    return again(0)

};

File: 0473-matchsticks-to-square/0473-matchsticks-to-square.py
from typing import List
class Solution:
    def makesquare(self, matchsticks: List[int]) -> bool:
        total_sum = sum(matchsticks)
        if total_sum % 4 != 0:
            return False
        target = total_sum // 4

        sides = [0] * 4
        matchsticks.sort(reverse=True)

        def backtrack(index): # 1,2
            if index == len(matchsticks):
                return True
            for i in range(4): # 0,1
                if sides[i] + matchsticks[index] <= target:
                    sides[i] += matchsticks[index]
                    if backtrack(index + 1):
                        return True
                    sides[i] -= matchsticks[index]
            return False

        return backtrack(0)
